rkt/containers.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2014 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/rocket/pkg/lock"
+)
+
+type container struct {
+	*lock.DirLock
+	uuid       string
+	isExited   bool
+	isGarbage  bool
+	isDeleting bool
+}
+
+type includeMask byte
+
+const (
+	includeContainersDir includeMask = 1 << iota
+	includeGarbageDir
+)
+
+var (
+	skipUUID = errors.New("skip this container")
+)
+
+// withContainerHandles iterates over the included containers calling function f for every container.
+func walkContainers(include includeMask, f func(string, *container, error) error) error {
+	ls, err := getContainers(include)
+	if err != nil {
+		return fmt.Errorf("failed to get containers: %v", err)
+	}
+
+	for _, uuid := range ls {
+		c, err := getContainer(uuid)
+
+		// omit garbage containers if garbage wasn't included
+		// this is to cover a race between getContainers finding the uuids and `rkt gc` renaming a container
+		if err != nil && c.isGarbage && include&includeGarbageDir == 0 {
```

Comment:
if `err != nil`, wouldn't `c == null`? And then `c.isGarbage` will blow up. Should this be `if err != nil || (c.isGarbage && include&includeGarbageDir == 0)`?